### PR TITLE
BUGZ-863: wait for objects to be added to physics before completing SafeLanding

### PR DESCRIPTION
**Pull request by AndrewMeadows**
https://highfidelity.atlassian.net/browse/BUGZ-863

This PR fixes a problem where: on arrival in a domain the physics simulation would be enabled before the floor was known to it, which could allow `MyAvatar` to fall through.

Also related: BUGZ-942 : sometimes SafeLanding never completes for model with bad collisionURL
https://highfidelity.atlassian.net/browse/BUGZ-942

### CODE REVIEW

**Comment by AndrewMeadows** [comment]
Yes, please review this PR, but don't bother setting DEV_APPROVED.  It can only be approved by committee.

File: interface/src/avatar/OtherAvatar.cpp [hunk 1]
```diff
@@ -201,6 +201,7 @@ void OtherAvatar::computeShapeLOD() {
         break;
     case workload::Region::UNKNOWN:
     case workload::Region::INVALID:
+    case workload::Region::R4:
```
Comment (AndrewMeadows) [inline interface/src/avatar/OtherAvatar.cpp]: I am salvaging some work from previously discarded PR-15707 where `workload::Region::R4` was added to disambiguate between objects that are "known to workload but not yet categorized" verses "known to workload and successfully categorized as outside `workload::Region::R3`".  Previously both of those categories were lumped into `workload::Region::UNKNOWN` because we didn't think the difference mattered.

File: interface/src/octree/SafeLanding.cpp [hunk 2]
```diff
@@ -122,9 +122,6 @@ void SafeLanding::updateTracking() {
             if (isEntityPhysicsReady(entity) && isVisuallyReady) {
                 entityMapIter = _trackedEntities.erase(entityMapIter);
             } else {
-                if (!isVisuallyReady) {
-                    entity->requestRenderUpdate();
-                }
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: In the last committee meeting Sam Gondelman pointed out: this `entity->requestRenderUpdate()` is not necessary.

File: interface/src/octree/SafeLanding.cpp [hunk 3]
```diff
@@ -203,8 +200,11 @@ bool SafeLanding::isEntityPhysicsReady(const EntityItemPointer& entity) {
             if (hasAABox && downloadedCollisionTypes.count(modelEntity->getShapeType()) != 0) {
                 auto space = _entityTreeRenderer->getWorkloadSpace();
                 uint8_t region = space ? space->getRegion(entity->getSpaceIndex()) : (uint8_t)workload::Region::INVALID;
-                bool shouldBePhysical = region < workload::Region::R3 && entity->shouldBePhysical();
-                return (!shouldBePhysical || entity->isInPhysicsSimulation() || modelEntity->computeShapeFailedToLoad());
+                bool definitelyNotPhysical = (region > workload::Region::R2 && region < workload::Region::UNKNOWN) ||
+                    !entity->shouldBePhysical() ||
+                    modelEntity->unableToLoadCollisionShape();
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: Here is where the differentiation between `workload::Region::UNKNOWN` (not yet categorized) and `workload::Region::R4` is important.  In general, we can't conclude as to whether the entity is `definitelyPhysical` or `definitelyNotPhysical` until the workload has properly categorized it.

File: interface/src/octree/SafeLanding.cpp (hunk 2 above)
Comment (SamGondelman) [reply interface/src/octree/SafeLanding.cpp]: https://highfidelity.atlassian.net/browse/BUGZ-902

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 4]
```diff
@@ -282,27 +282,26 @@ bool RenderableModelEntityItem::findDetailedParabolaIntersection(const glm::vec3
 }
 
 void RenderableModelEntityItem::fetchCollisionGeometryResource() {
-    _compoundShapeResource = DependencyManager::get<ModelCache>()->getCollisionGeometryResource(getCollisionShapeURL());
+    _collisionGeometryResource = DependencyManager::get<ModelCache>()->getCollisionGeometryResource(getCollisionShapeURL());
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: This is a name change to properly reflect we are talking about the collision shape, which could be using the "compound" feature but might not (e.g. it could be using the exact visible mesh for its collision shape).

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 5]
```diff
@@ -960,6 +959,10 @@ QStringList RenderableModelEntityItem::getJointNames() const {
     return result;
 }
 
+QString RenderableModelEntityItem::getCollisionShapeURL() const {
+    return getShapeType() == SHAPE_TYPE_COMPOUND ? getCompoundShapeURL() : getModelURL();
+}
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: This was moved out of `ModelEntityItem` and into `RenderModelEntityItem`, which is the only context where it is used.

File: libraries/physics/src/PhysicalEntitySimulation.cpp [hunk 6]
```diff
@@ -330,6 +330,18 @@ void PhysicalEntitySimulation::buildMotionStatesForEntitiesThatNeedThem() {
...
+        if (region == workload::Region::UNKNOWN) {
+            // the workload hasn't categorized it yet --> skip for later
+            ++entityItr;
+            continue;
+        }
+        if (region > workload::Region::R2) {
+            // not in physical zone --> remove from list
+            entityItr = _entitiesToAddToPhysics.erase(entityItr);
+            continue;
+        }
```
Comment (AndrewMeadows) [inline libraries/physics/src/PhysicalEntitySimulation.cpp]: Double-check the workload region and decide to skip (not yet categorized) or remove (entity isn't in the physics simulation zone after all) as the case may be.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20241/HighFidelity-Beta-PR15880-6f7c029.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20241/HighFidelity-Quest-Beta-PR15880-6f7c029.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20241/HighFidelity-Beta-PR15880-6f7c029.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20241/HighFidelity-Beta-Interface-PR15880-6f7c029.dmg

File: interface/src/octree/SafeLanding.cpp (hunk 2 above)
Comment (SamGondelman) [reply interface/src/octree/SafeLanding.cpp]: Actually, I would recommend doing this in a separate PR.  I don't think the _visuallyReady flag is updated properly for all entity types, and this could cause problems.

For example, right now we only set it for models, particles, and zones.  A quick glance at RenderableParticleEffectEntityItem doesn't convince me that it's asking for enough render updates.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20241/HighFidelity-Beta-PR15880-6f7c029.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20241/HighFidelity-Beta-Interface-PR15880-6f7c029.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20271/HighFidelity-Beta-PR15880-b7cab54.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20271/HighFidelity-Beta-Interface-PR15880-b7cab54.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20271/HighFidelity-Beta-PR15880-b7cab54.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20271/HighFidelity-Quest-Beta-PR15880-b7cab54.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20271/HighFidelity-Beta-PR15880-b7cab54.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20271/HighFidelity-Beta-Interface-PR15880-b7cab54.exe

File: libraries/workload/src/workload/RegionState.cpp [hunk 7]
```diff
@@ -28,7 +28,9 @@ void RegionState::run(const workload::WorkloadContextPointer& renderContext, con
     // ...
     // inputs[2N] = vector of ids exiting region N
     // inputs[2N + 1] = vector of ids entering region N
-    assert(inputs.size() == 2 * Region::UNKNOWN);
+    //
+    // But we only pass inputs for R1 through R3
+    assert(inputs.size() == 2 * (int32_t)(Region::R3 + 1));
```
Comment (sabrina-shanman) [inline libraries/workload/src/workload/RegionState.cpp]: This assertion doesn't match the comment.

File: interface/src/Application.cpp [hunk 8]
```diff
@@ -6952,13 +6952,15 @@ int Application::sendNackPackets() {
...
-
     const bool isModifiedQuery = !_physicsEnabled;
     if (isModifiedQuery) {
+        if (!_settingsLoaded) {
+            return; // bail early if settings are not loaded
+        }
+
+        if (!_octreeProcessor.safeLandingIsComplete() && !_octreeProcessor.safeLandingIsActive()) {
+            return;
+        }
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: This fixes BUGZ-942. I was seeing the entity-server send the "ok I've sent all sequence numbers" message before the interface-client had even started the SafeLanding logic.  Somehow this would could cause SafeLanding to get confused an never notice that the bad Resource had failed to load.  The fix is in but I'm still trying to understand it better.

File: libraries/workload/src/workload/RegionState.cpp (hunk 7 above)
Comment (AndrewMeadows) [reply libraries/workload/src/workload/RegionState.cpp]: Actually it does:  `R0 = 0` (index starts at zero) and `R3 = 2`, and the inputs for R0 to R3 (at two inputs per region) would have `size = 2*(R3+1) = 2*(2+1) = 6`.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20273/HighFidelity-Beta-PR15880-f0fa5d0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20273/HighFidelity-Beta-Interface-PR15880-f0fa5d0.dmg

File: libraries/workload/src/workload/RegionState.cpp (hunk 7 above)
Comment (sabrina-shanman) [reply libraries/workload/src/workload/RegionState.cpp]: Ah, I see, I was mixing up the indexing in the comments with the size in the assert. Makes sense!

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20273/HighFidelity-Beta-PR15880-f0fa5d0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20273/HighFidelity-Beta-Interface-PR15880-f0fa5d0.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20273/HighFidelity-Beta-PR15880-f0fa5d0.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20273/HighFidelity-Quest-Beta-PR15880-f0fa5d0.apk)

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp (hunk 5 above)
Comment (ZappoMan) [reply libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Is it possible to have other shape types here?
How is collisionless handled in this case? wouldn't a NULL collision shape URL be more logical in that case?

Comment (ZappoMan) [reply libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Why did you move this from the base class?

File: libraries/entities/src/ModelEntityItem.cpp [hunk 9]
```diff
@@ -590,10 +590,6 @@ QString ModelEntityItem::getCompoundShapeURL() const {
     return _compoundShapeURL.get();
 }
 
-QString ModelEntityItem::getCollisionShapeURL() const {
```
Comment (ZappoMan) [inline libraries/entities/src/ModelEntityItem.cpp]: why was this moved into renderable? isn't it meaningful for other code to be able to know the URL even if not rendering?

File: libraries/workload/src/workload/RegionState.h [hunk 10]
```diff
@@ -54,7 +54,7 @@ namespace workload {
         using JobModel = workload::Job::ModelI<RegionState, Inputs, Config>;
 
         RegionState() {
-            _state.resize(Region::UNKNOWN);
+            _state.resize(Region::R3 + 1);
```
Comment (ZappoMan) [inline libraries/workload/src/workload/RegionState.h]: this feels like it could use a comment, especially since Region has several other items in it

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20298/HighFidelity-Beta-PR15880-9a6b42e.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20298/HighFidelity-Quest-Beta-PR15880-9a6b42e.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20298/HighFidelity-Beta-PR15880-9a6b42e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20298/HighFidelity-Beta-Interface-PR15880-9a6b42e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20298/HighFidelity-Beta-PR15880-9a6b42e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20298/HighFidelity-Beta-Interface-PR15880-9a6b42e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20310/HighFidelity-Beta-PR15880-1a5a925.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20310/HighFidelity-Beta-Interface-PR15880-1a5a925.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20310/HighFidelity-Beta-PR15880-1a5a925.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20310/HighFidelity-Quest-Beta-PR15880-1a5a925.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20310/HighFidelity-Beta-PR15880-1a5a925.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20310/HighFidelity-Beta-Interface-PR15880-1a5a925.exe

File: libraries/workload/src/workload/RegionTracker.cpp [hunk 11]
```diff
@@ -33,15 +33,15 @@ void RegionTracker::run(const WorkloadContextPointer& context, Outputs& outputs)
         //Changes changes;
         space->categorizeAndGetChanges(outChanges);
 
-        // use exit/enter lists for each region less than Region::UNKNOWN
+        // use exit/enter lists for each region less than Region::R4
         outRegionChanges.resize(2 * (workload::Region::NUM_CLASSIFICATIONS - 1));
```
Comment (sabrina-shanman) [inline libraries/workload/src/workload/RegionTracker.cpp]: Not a blocker: Perhaps consider changing this to use the new NUM_REGIONS_TRACKED to make this more explicit.

File: libraries/workload/src/workload/Region.h [hunk 12]
```diff
@@ -18,11 +18,12 @@ class Region {
...
-        INVALID,
+        R1 = 0,   // R1 = in physics simulation and client will bid for simulation ownership
+        R2,       // R2 = in physics simulation but client prefers to NOT have simulation ownership
+        R3,       // R3 = are NOT in physics simulation but yes kinematically animated when velocities are non-zero
+        R4,       // R4 = known to workload but outside R3, not in physics, not animated if moving
+        UNKNOWN,  // UNKNOWN = known to workload but unsorted
+        INVALID,  // INVALID = not known to workload
     };
 
     static const uint8_t NUM_CLASSIFICATIONS = 4;
```
Comment (sabrina-shanman) [inline libraries/workload/src/workload/Region.h]: While I'm talking about names: computeTransitionIndex and NUM_TRANSITIONS appear to be unused, and NUM_CLASSIFICATIONS seems similar to NUM_REGIONS_TRACKED..

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20336/HighFidelity-Beta-PR15880-ba89788.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20336/HighFidelity-Quest-Beta-PR15880-ba89788.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20336/HighFidelity-Beta-PR15880-ba89788.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20336/HighFidelity-Beta-Interface-PR15880-ba89788.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20342/HighFidelity-Beta-PR15880-f165184.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20342/HighFidelity-Quest-Beta-PR15880-f165184.apk)

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20344/HighFidelity-Beta-PR15880-0d943ca.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20344/HighFidelity-Quest-Beta-PR15880-0d943ca.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20344/HighFidelity-Beta-PR15880-0d943ca.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20344/HighFidelity-Beta-Interface-PR15880-0d943ca.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20344/HighFidelity-Beta-PR15880-0d943ca.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20344/HighFidelity-Beta-Interface-PR15880-0d943ca.exe

File: libraries/workload/src/workload/RegionState.h [hunk 13]
```diff
@@ -53,8 +53,11 @@ namespace workload {
         using Inputs = IndexVectors;
         using JobModel = workload::Job::ModelI<RegionState, Inputs, Config>;
 
+        // we track Proxies in regions R1 through R3
+        const uint32_t NUM_REGIONS_TRACKED = uint32_t(Region::R3 - Region::R1 + 1);
```
Comment (ZappoMan) [inline libraries/workload/src/workload/RegionState.h]: this looks so similar to `Region::NUM_TRACKED_REGIONS`... and they even have the same value.

Is there a reason they are almost identical but not identical? 

Can we either:
- pick different names because they are very different
or 
- use the same value

Comment (AndrewMeadows) [reply libraries/workload/src/workload/RegionState.h]: Ah yes, thanks for spotting that.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20353/HighFidelity-Beta-PR15880-590876e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20353/HighFidelity-Beta-Interface-PR15880-590876e.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20353/HighFidelity-Beta-PR15880-590876e.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20353/HighFidelity-Quest-Beta-PR15880-590876e.apk)

**Comment by AndrewMeadows** [comment]
Jenkins, rebuild this please.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20355/HighFidelity-Beta-PR15880-590876e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20355/HighFidelity-Beta-Interface-PR15880-590876e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20355/HighFidelity-Beta-PR15880-590876e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20355/HighFidelity-Beta-Interface-PR15880-590876e.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20355/HighFidelity-Beta-PR15880-590876e.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20355/HighFidelity-Quest-Beta-PR15880-590876e.apk)
